### PR TITLE
[Build] Add srt_empty extra group for device-agnostic install

### PR DESCRIPTION
**Pull request by yixiaodapeng**
## Motivation

SGLang's `runtime_common` in `pyproject_other.toml` includes `torchao` and `timm`, which pull in a specific torch version and conflict with vendor-specific PyTorch builds (`torch_npu`, `torch_musa`, etc.). This prevents OOT plugins from using SGLang as a pure-Python base on non-NVIDIA platforms.

vLLM solved the same problem with `VLLM_TARGET_DEVICE=empty` ([PR #4773](https://github.com/vllm-project/vllm/pull/4773)). SGLang uses declarative `pyproject.toml` instead of dynamic `setup.py`, so we add an extra group rather than a build-time env var.

## Changes

| File | Change |
|------|--------|
| `python/pyproject_other.toml` | Split `runtime_common` into `runtime_base` (torch-free, ~48 packages) + `runtime_common` (backward-compatible = runtime_base + torchao + timm). Add `srt_empty` extra group. |
| `python/sglang/srt/server_args.py` | `try/except` for `FLA_CHUNK_SIZE` import (avoids top-level `import triton`) |
| `python/sglang/srt/managers/scheduler.py` | `try/except` for `torch.cuda.Stream` import |

**Backward-compatible**: `pip install -e ".[runtime_common]"` installs exactly the same packages as before.

## Usage

```bash
cp pyproject_other.toml pyproject.toml
pip install -e ".[srt_empty]"
```

This is the same `cp` workflow already used by AMD/MUSA/HPU platforms — `srt_empty` simply adds another option that skips torch-pulling dependencies.

## Testing

Verified on NVIDIA H20 (8×144GB) with [sglang-plugin-FL](https://github.com/flagos-ai/sglang-plugin-FL) + [FlagGems](https://github.com/FlagOpen/FlagGems) (Triton attention backend, no sgl_kernel/flashinfer dependency):

| Model | Type | TP | Test | Result |
|-------|------|----|------|--------|
| Qwen3.6-27B | Dense + FLA | 4 | Offline inference (text + VL) | ✅ Pass |
| Qwen3.6-35B-A3B | MoE (256 experts) | 4 | Offline inference (text + VL) | ✅ Pass |
| Qwen3.6-27B | Dense + FLA | 4 | 16-request concurrent | ✅ Pass |
| Qwen3.6-35B-A3B | MoE (256 experts) | 4 | 16-request concurrent | ✅ Pass |

All ops dispatch to [FlagGems](https://github.com/FlagOpen/FlagGems) Triton or reference implementations — zero dependency on sgl_kernel or flashinfer.

## Context

This PR enables the [sglang-plugin-FL](https://github.com/flagos-ai/sglang-plugin-FL) OOT plugin ecosystem, which provides multi-chip inference support for SGLang across diverse hardware (Ascend, MUSA, etc.) using [FlagGems](https://github.com/FlagOpen/FlagGems) (unified Triton operator library) and [FlagCX](https://github.com/flagos-ai/FlagCX) (unified communication library). The plugin uses SGLang's hook system (introduced in [PR #21388](https://github.com/sgl-project/sglang/pull/21388)) to intercept fused ops and route them through platform-agnostic backends.

## Affiliation

Beijing Academy of Artificial Intelligence (BAAI) — FlagOS / AI Framework Team

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #30983571819](https://github.com/sgl-project/sglang/actions/runs/30983571819)
Latest PR Test (Extra): :x: [Run #30983571449](https://github.com/sgl-project/sglang/actions/runs/30983571449)

### CODE REVIEW

File: python/pyproject_other.toml [hunk 1]
```diff
@@ -17,7 +17,9 @@ classifiers = [
 dependencies = ["aiohttp", "IPython", "numpy", "requests", "setproctitle", "tqdm"]
 
 [project.optional-dependencies]
-runtime_common = [
+# runtime_base: all pure-Python packages from runtime_common, WITHOUT packages
+# that pull in torch (torchao, timm). Used by srt_empty for device-agnostic install.
+runtime_base = [
```
Comment (alexnails) [inline python/pyproject_other.toml:23]: [Agent dive on runtime_base]

Resolving it for `linux/x86_64` / py3.10:

```
torch==2.13.0
triton==3.7.1
nvidia-cublas, nvidia-cudnn-cu13, nvidia-nccl-cu13, ...  (15 nvidia-* wheels)
-> 163 packages total
```

Three entries declare torch unconditionally:

| package | requirement |
|---|---|
| `xgrammar==0.2.1` | `torch>=1.10.0` **and** `triton; Linux and x86_64` |
| `compressed-tensors` (unpinned) | `torch>=2.10.0` in the current release |
| `outlines==0.1.11` | `torch` |

Removing just those three gives **128 packages, zero torch/triton/nvidia**.

The `triton` dep installs upstream NVIDIA triton on Linux x86_64, which collides with vendor triton forks (triton-ascend et al.) — the exact conflict class this PR exists to avoid. And `compressed-tensors`' `torch>=2.10.0` will actively *upgrade over* a vendor torch rather than being satisfied by it.

Moving those three into `runtime_common` fixes it, with one trade-off worth stating explicitly: the default `--grammar-backend` is `xgrammar` (`server_args.py:4996`) and `xgrammar_backend.py:22` imports it at module scope, so an `srt_empty` install has no working structured output until the user installs a backend themselves. That seems fine for an "empty" target, but it should be documented, and the PR description's "no torch dependency chain" claim needs updating either way.

File: python/pyproject_other.toml [hunk 2]
```diff
@@ -17,7 +17,9 @@ classifiers = [
 dependencies = ["aiohttp", "IPython", "numpy", "requests", "setproctitle", "tqdm"]
 
 [project.optional-dependencies]
-runtime_common = [
+# runtime_base: all pure-Python packages from runtime_common, WITHOUT packages
+# that pull in torch (torchao, timm). Used by srt_empty for device-agnostic install.
```
Comment (alexnails) [inline python/pyproject_other.toml]: "torch-free" is probably more apt description here

File: python/pyproject_other.toml [hunk 3]
```diff
@@ -59,15 +61,26 @@ runtime_common = [
...
+  "sglang[runtime_base]",
+  "timm==1.0.16",
+  "torchao==0.9.0",
+]
+
+# srt_empty: device-agnostic install — pure Python packages only, no torch dependency chain.
+# Enables OOT plugins (e.g. sglang-plugin-FL) to install sglang without conflicting with
+# vendor-specific PyTorch builds (torch_npu, torch_musa, etc.).
+# Usage: cp pyproject_other.toml pyproject.toml && pip install -e ".[srt_empty]"
+srt_empty = ["sglang[runtime_base]"]
```
Comment (alexnails) [inline python/pyproject_other.toml:85]: Ask: let's have a pre-commit hook or registered unit test that resolves `runtime_base` and asserts `torch` isn't in the graph. A `dev_empty = ["sglang[srt_empty]", "sglang[test]"]` aggregate would also let CI smoke-test the target directly.

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -35,7 +35,10 @@
 import setproctitle
 import torch
 import torch.distributed
-from torch.cuda import Stream as CudaStream
+try:
+    from torch.cuda import Stream as CudaStream
+except ImportError:
+    CudaStream = None  # type: ignore[misc,assignment]
```
Comment (alexnails) [inline python/sglang/srt/managers/scheduler.py]: CudaStream with none just tells the type checker `copy_stream: None`, which is why the `type: ignore[misc,assignment]` became necessary.

We can do something like:

```python
from typing import TYPE_CHECKING, Any, Deque, ...

if TYPE_CHECKING:
    from torch.cuda import Stream as CudaStream
    from torch.cuda import StreamContext as CudaStreamContext
```

That also lets us delete the `CudaStreamContext = nullcontext` half of the `is_mps()` branch at lines 283-287, which exists purely to annotate type.

On the TODO side of things: the multi hardware plugin will need a `DeviceStream` / `DeviceStreamContext` alias in the platforms package next to `DeviceMixin`

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -32,7 +32,10 @@
 from typing import Any, Callable, Dict, List, Literal, Optional, Union
 
 from sglang.jit_kernel.kv_canary.consts import RealKvHashMode
-from sglang.kernels.ops.attention.fla.chunk_delta_h import CHUNK_SIZE as FLA_CHUNK_SIZE
+try:
+    from sglang.kernels.ops.attention.fla.chunk_delta_h import CHUNK_SIZE as FLA_CHUNK_SIZE
+except ImportError:
+    FLA_CHUNK_SIZE = 64  # Default from fla/chunk_delta_h.py; avoids import triton at top level
```
Comment (alexnails) [inline python/sglang/srt/server_args.py]: With this change there, can now be drift with FLA_CHUNK_SIZE?

File: python/sglang/srt/managers/scheduler.py [hunk 6]
```diff
@@ -35,7 +35,10 @@
...
 from torch.distributed import barrier
 
 from sglang.kernels.ops.mamba.triton_ops import (
```
Comment (alexnails) [inline python/sglang/srt/managers/scheduler.py]: still pull in triton here at top level

**Review by alexnails — COMMENTED** [review]
Specific things I believe we need to fix in this PR or in following PRs:

1. runtime_base still pulls torch, triton and the CUDA stack. Agent started resolving the 46-package list for linux/x86_64 / py3.10 with uv pip compile gives torch==2.13.0, triton==3.7.1, and 15 nvidia-* wheels (163 packages total).

2. CudaStream / CudaStreamContext are only ever used as annotations and are never evaluated at runtime. I proposed a solution in my comment.

3. FLA_CHUNK_SIZE affects mamba and there could now be drift

File: python/pyproject_other.toml (hunk 1 above)
Comment (yixiaodapeng) [reply python/pyproject_other.toml:23]: Done. Moved compressed-tensors, outlines, and xgrammar to runtime_common.

File: python/pyproject_other.toml (hunk 2 above)
Comment (yixiaodapeng) [reply python/pyproject_other.toml]: Done

File: python/pyproject_other.toml (hunk 3 above)
Comment (yixiaodapeng) [reply python/pyproject_other.toml:85]: Added test/registered/core/test_srt_empty_deps.py — a pytest that parses runtime_base and asserts known torch-pulling packages aren't present. Full transitive resolution (uv pip compile) can be added as a CI step in a follow-up.

File: python/sglang/srt/managers/scheduler.py (hunk 4 above)
Comment (yixiaodapeng) [reply python/sglang/srt/managers/scheduler.py]: Switched to if TYPE_CHECKING pattern.

File: python/sglang/srt/managers/scheduler.py (hunk 6 above)
Comment (yixiaodapeng) [reply python/sglang/srt/managers/scheduler.py]: Wrapped in try/except with None fallback. Added a guard in init_mamba_backend() to skip initialization when unavailable.

File: python/sglang/srt/server_args.py (hunk 5 above)
Comment (yixiaodapeng) [reply python/sglang/srt/server_args.py]: Moved the import into mamba_cache_chunk_size property as a lazy import with a # Must match ... comment on the fallback. 
This reduces top-level exposure but doesn't fully eliminate drift risk — the hardcoded 64 is still there as a last-resort fallback. 
Extracting to a shared triton-free consts file would be the clean fix, but it touches upstream file structure. Do you think that's worth doing in this PR, or acceptable as a follow-up?

**Review by alexnails — APPROVED** [review]

**Comment by alexnails** [comment]
@yixiaodapeng please look at NPU CI Failures, there seems to be some issue?

**Comment by yixiaodapeng** [comment]
> @yixiaodapeng please look at NPU CI Failures, there seems to be some issue?

Okay
